source/codegen/generate_service.py

```diff
@@ -25,11 +25,44 @@ def generate_service_file(metadata, template_file_name, generated_file_suffix, g
   f.write(template.render(data=metadata))
   f.close()
 
+
+class AttributeValueExpander:
+  '''Wraps an _attribute_type_map of:
+  
+  group -> type -> attributes
+
+  and implements the metadata_mutations to add_attribute_values_enums and expand_attribute_function_value_param.
+  '''
+  def __init__(self, metadata):
+    self._metadata = metadata
+    self._attribute_type_map = {}
+
+    for group_name, attributes in common_helpers.get_attribute_groups(metadata).items():
+      attribute_enums_by_type = common_helpers.get_attribute_enums_by_type(attributes)
+      metadata_mutation.add_attribute_values_enums(metadata["enums"], attribute_enums_by_type, group_name)
+      self._attribute_type_map[group_name] = attribute_enums_by_type
+
+  def get_used_attribute(self, parameters):
+    param_types = (param["grpc_type"] for param in parameters)
```

Comment:
Ah, its not looking for value parameter. It is actually looking for parameter which accepts attribute id and just assumes there would be another parameter for attribute value that needs to be mutated. Sneaky! :)

The reason why this is working for MI drivers is because all attribute params for MI drivers are marked as `ViAttr` type and then `grpc_type` for those is updated [here](https://github.com/ni/grpc-device/blob/c01e3769891e3443fe0add6e6947612a7b6867fb/source/codegen/metadata_mutation.py#L106).

Reply:
Thanks for the feedback. The name change you proposed below made this a little clearer. And I added a comment. I'll have to make another pass on this for the data-type-specific attributes. SO, hopefully I can make it less-not-more confusing.